### PR TITLE
[RFC4_dev] Use Win32 Unicode APIs and expect all strings to be UTF-8 (fixes #1765)

### LINKED ISSUE #1765: PROJ_LIB environment variable non ANSI characters in the path
## Expected behavior and actual behavior.

PROJ failed to read PROJ_LIB environment variable when the file path encoding
is different from the system local 

## Steps to reproduce the problem.
1. set the  PROJ_LIB environment variable to something different from your current locale (for example c:\temp\档案文件\proj )

2. add a egm96_15.gtx file to the above directory 

3.  transform failed on trying to open the GTX file
 

## Operating system

Windows 10

## PROJ version and provenance

PROJ 4.9.3

### PR DESCRIPTION
**Pull request by rouault**
Not completely RFC4 specific but proper unicode support for Windows is now really needed given all the file manipulations we do.
continuation of PR #1826, starting with commit f6c4241

### CODE REVIEW

**Review by kbevers — COMMENTED** [review]
The bulk of this PR is the unicode stuff which isn't something I feel comfortable reviewing. I hope someone else can review this.

I appreciate that a replacement for the `proj_api.h` fileAPI has been added to `proj.h`. On a functional level I am also happy to see unicode support added!

**Comment by rouault** [comment]
> The bulk of this PR is the unicode stuff

It is largely inspired from what is used in GDAL: UTF-8 <--> widechar conversions and use of Win32 Widechar File APIs

**Comment by rouault** [comment]
@hobu Want to give that a review ?

File: src/grids.cpp [hunk 1]
```diff
@@ -1790,6 +1838,11 @@ std::unique_ptr<NTv2GridSet> NTv2GridSet::open(PJ_CONTEXT *ctx,
         pj_ctx_set_errno(ctx, PJD_ERR_FAILED_TO_LOAD_GRID);
         return nullptr;
     }
+    if (memcmp(header + 56, "SECONDS", 7) != 0) {
```
Comment (hobu) [inline src/grids.cpp:1841]: This looks like a magic number of some kind. A comment about what it is or why it is here would be helpful.

File: src/grids.cpp [hunk 2]
```diff
@@ -2644,4 +2773,612 @@ ListOfGenericGrids proj_generic_grid_init(PJ *P, const char *gridkey) {
...
+
+    return getListOfGridSets(P->ctx, grids);
+}
+
+// ---------------------------------------------------------------------------
+
+typedef struct { pj_int32 lam, phi; } ILP;
+
+static PJ_LP pj_hgrid_interpolate(PJ_LP t, const HorizontalShiftGrid *grid,
+                                  bool compensateNTConvention) {
```
Comment (hobu) [inline src/grids.cpp:2846]: A description of what this interpolation method is, how it was constructed, a reference to a paper? that states this is a good method to use, etc., would be helpful for posterity.

File: src/grids.cpp [hunk 3]
```diff
@@ -2644,4 +2773,612 @@ ListOfGenericGrids proj_generic_grid_init(PJ *P, const char *gridkey) {
...
+}
+
+// ---------------------------------------------------------------------------
+
+bool pj_bilinear_interpolation_three_samples(const GenericShiftGrid *grid,
+                                             const PJ_LP &lp, int idx1,
+                                             int idx2, int idx3, double &v1,
+                                             double &v2, double &v3,
+                                             bool &must_retry) {
+    must_retry = false;
```
Comment (hobu) [inline src/grids.cpp:3285]: Docs on this method, history, etc., needed.

File: src/proj_internal.h [hunk 4]
```diff
@@ -677,6 +677,31 @@ struct projNetworkCallbacksAndData
     void* user_data = nullptr;
 };
 
+struct projGridChunkCache
+{
+    bool enabled = true;
+    std::string filename{};
+    long long max_size = 100 * 1024 * 1024;
```
Comment (hobu) [inline src/proj_internal.h:684]: We had previously discussed a cache size of 300mb...

File: test/cli/Makefile.am [hunk 5]
```diff
@@ -36,24 +36,24 @@ test83-check:
 	$(TEST83) $(PROJEXE)
 
 testvarious-check:
-	PROJ_LIB=$(PROJ_LIB) $(TESTVARIOUS) $(CS2CSEXE)
+	PROJ_IGNORE_USER_WRITABLE_DIRECTORY=YES PROJ_LIB=$(PROJ_LIB) $(TESTVARIOUS) $(CS2CSEXE)
```
Comment (hobu) [inline test/cli/Makefile.am:39]: I worry that ``PROJ_IGNORE_USER_WRITABLE_DIRECTORY`` isn't a strong enough name. You are using it negated in this incantination. Does this variable mean to not write into user directories? Not read any data from user directories? It isn't quite clear.

File: test/unit/gie_self_tests.cpp [hunk 6]
```diff
@@ -348,7 +348,9 @@ TEST(gie, info_functions) {
 
     /* proj_info()                                                            */
     /* this one is difficult to test, since the output changes with the setup */
+    putenv(const_cast<char *>("PROJ_IGNORE_USER_WRITABLE_DIRECTORY="));
```
Comment (hobu) [inline test/unit/gie_self_tests.cpp:351]: Again, ``PROJ_IGNORE_USER_WRITABLE_DIRECTORY`` has spread everywhere. Does this mean disable the cache or disable storing it in per-user locations like LOCALDATA?

**Comment by rouault** [comment]
@hobu Parts of your comments are valid, but you're actually reviewing old commits. The one specifics of this PR start with https://github.com/OSGeo/PROJ/commit/f6c424165264eec06af2504800ebfdd5d701efae . In order to avoid painful rebase, I queue review commits in the most recent pull request of this rfc4_dev series, currently https://github.com/OSGeo/PROJ/pull/1857

File: src/grids.cpp (hunk 1 above)
Comment (rouault) [reply src/grids.cpp:1841]: symbolic name added

File: src/grids.cpp (hunk 2 above)
Comment (rouault) [reply src/grids.cpp:2846]: comment added

File: src/grids.cpp (hunk 3 above)
Comment (rouault) [reply src/grids.cpp:3285]: comment added

File: src/proj_internal.h (hunk 4 above)
Comment (rouault) [reply src/proj_internal.h:684]: indeed. Was changed in proj.ini & doc, but not here. Fixed

File: test/cli/Makefile.am (hunk 5 above)
Comment (rouault) [reply test/cli/Makefile.am:39]: changed to PROJ_SKIP_READ_USER_WRITABLE_DIRECTORY

File: test/unit/gie_self_tests.cpp (hunk 6 above)
Comment (rouault) [reply test/unit/gie_self_tests.cpp:351]: changed to PROJ_SKIP_READ_USER_WRITABLE_DIRECTORY

**Comment by rouault** [comment]
Unless someone has further comments, I'll merge that tomorrow

File: scripts/reformat_cpp.sh [hunk 7]
```diff
@@ -18,7 +18,8 @@ TOPDIR="$SCRIPT_DIR/.."
 for i in "$TOPDIR"/include/proj/*.hpp "$TOPDIR"/include/proj/internal/*.hpp \
          "$TOPDIR"/src/iso19111/*.cpp "$TOPDIR"/test/unit/*.cpp "$TOPDIR"/src/apps/projinfo.cpp \
          "$TOPDIR"/src/tracing.cpp "$TOPDIR"/src/grids.hpp "$TOPDIR"/src/grids.cpp \
-         "$TOPDIR"/src/filemanager.hpp "$TOPDIR"/src/filemanager.cpp ; do
+         "$TOPDIR"/src/filemanager.hpp "$TOPDIR"/src/filemanager.cpp \
+         "$TOPDIR"/src/sqlite3.hpp "$TOPDIR"/src/sqlite3.cpp ; do
```
Comment (mwtoews) [inline scripts/reformat_cpp.sh:22]: Shouldn't this be `"$TOPDIR"/src/sqlite3_utils.hpp "$TOPDIR"/src/sqlite3_utils.cpp`?

Comment (rouault) [reply scripts/reformat_cpp.sh:22]: Good catch. Updated

**Pull request merged**